### PR TITLE
Create Feature Flag that is Protected by a Test Team Membership

### PR DESCRIPTION
**Pull request by sarahzinger**
## Links
* Remix link: https://sponge-credit.glitch.me/
* Issue-tracker link: https://app.clubhouse.io/glitch/story/6208/create-feature-flag
* Specs/doc links: https://docs.google.com/document/d/1a4s0HSkkwq3nuNsX4GqrC7sVrA7AOJ255psYlal9dFc/edit?ts=5e052df8

## Changes:
Hopefully nothing very noticeable :) 
My intention here was to:
- add a new Optimizely attribute that we can use to build a new audience for a new feature flag of Pufferfish
- protect that flag so that it only works if Optimizely says so AND if we enable it on the /secret page. For our other flags we enable it if optimizely says so OR if it's enabled on the /secret page.
- The general gist here is that we want Folks apart of the Test Team to be able to toggle the new code whenever they like on /secret, but that the general public can't get a preview of our code right now. 

## How To Test:
* ~I'm not sure :( I don't see anything different, but my guess is we have to deploy first to see the changes in Optimizely. Maybe @GregWeil you have thoughts, I know you have a lot expertise in this area? Maybe I can run this code in a prod env or something to get Optimizely to pick up the changes?~
* Add this key to your .env file (https://github.com/FogCreek/Glitch-Community/pull/1151#pullrequestreview-339493595) or use the remix where I'll add it myself
* Check out this table, does it mesh with your experience?

| Case  | Expected Outcome | 
| ------------- | ------------- |
| In Testing Team  + haven't done anything on /secret   |  /settings does NOT say pufferfish |
| In Testing Team + Enabled the feature on /secret  | /settings says pufferfish (weird flash of 404 ok for now, getting fixed in a future pr) |
| In Testing Team + Disabled the feature on /secret  | settings does NOT say pufferfish |
| NOT In Testing Team + haven't done anything on /secret   |  /settings does NOT say pufferfish |
| NOT In Testing Team + Enabled the feature on /secret  | settings does NOT say pufferfish  |
| NOT In Testing Team + Disabled the feature on /secret  | settings does NOT say pufferfish |

## Feedback I'm looking for:
- Is there anyway to get the pufferfish feature flag without being a member of the glitch team (through Optimizely cookies or something?). I'm not sure if I need another manual check here in our code. 
- This feels like "A Lot" but my hope here is that if we ever need this additional protection in the future we can reuse this code, does that seem to be the case?
- I copied over a lot of stuff based on other Optimizely features and SSR and loosely followed where all these things were being used, but if I forgot something (or added something I don't need to in this case) feel free to shout!
- Does the /secret page make sense? I feel like we could simplify but it's starting to hurt my brain a bit so maybe we could do that in a future pr or something
- I moved the check for if a user is in the testing team to constants because if I left it in rollouts.js it created a circular dependency with current user, but maybe there's a better place? feel free to let me know if you think of somewhere else it should be

## To do before deploy
I could remove the changes made to /settings before deploying (although I don't think it's a big deal if we keep it?) just wanted to have it so folks could test with it more easily

### CODE REVIEW

**Comment by sarahzinger** [comment]
cc @FogCreek/pufferfish it's an early pass, but just as a heads up

File: src/state/current-user.js [hunk 1]
```diff
@@ -11,10 +11,10 @@ import { getStorage, readFromStorage, writeToStorage } from './local-storage';
 import { getAPIForToken } from './api'; // eslint-disable-line import/no-cycle
 import { appMounted } from './app-mounted';
 
+const TESTING_TEAM_ID = 3247;
```
Comment (keithk) [inline src/state/current-user.js]: Should we have this in our constants, instead of in a couple of places?

**Comment by keithk** [comment]
@sarahzinger You can create a new flag in optimizely here: https://app.optimizely.com/v2/projects/16762000107/features

And you can set it to be deployed to **development only** and then you can validate this locally

File: src/state/current-user.js [hunk 2]
```diff
@@ -39,6 +39,7 @@ function identifyUser(user) {
   }
   setCookie('hasLogin', user && user.login);
   setCookie('hasProjects', user && user.projects.length > 0);
+  setCookie('inTestingTeam', user && user.login && user.teams && user.teams.filter((t) => t.id === TESTING_TEAM_ID).length > 0);
```
[the remark this answers is not in the document]
Comment (sarahzinger) [reply src/state/current-user.js]: good call, I'll make a comment in the code

**Comment by sarahzinger** [comment]
thanks for the feedback @keithk @GregWeil I think I got it working 🙌 if y'all want to take another look. Also I updated the PR description if that helps

**Comment by GregWeil** [comment]
The only way to override what optimizely returns is to use `optimizelyClientInstance.setForcedVariation(experimentid, userid, variationid)`, with the ids pulled from the config. I don't think it's an issue, but I don't know how strict we have to be on this.

**Comment by sarahzinger** [comment]
Ok I tried using the forced variation call to see if I could "hack it" and it didn't enable the feature when it shouldn't. Just took a look in the docs here https://docs.developers.optimizely.com/full-stack/docs/use-forced-bucketing and it seems like this isn't supported for rollouts anyway. Feels like we're reasonably solid for this particular situation. 

Also broke out the most yucky bit into it's own function and wrote some tests cuz I was feeling guilty about it.

**Pull request merged**